### PR TITLE
Generate process metrics with `semconv` yaml

### PR DESCRIPTION
**Pull request by braydonk**
Supplants #160 

## Changes

This PR adds a yaml version of the process metrics and changes the process metrics documentation to be generated from the yaml instead of hand-written.

## Merge requirement checklist

* [x] [CONTRIBUTING.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md) guidelines followed.
* [x] [CHANGELOG.md](https://github.com/open-telemetry/semantic-conventions/blob/main/CHANGELOG.md) updated for non-trivial changes.
* [x] [schema-next.yaml](https://github.com/open-telemetry/semantic-conventions/blob/main/schema-next.yaml) updated with changes to existing conventions.

### CODE REVIEW

**Review by mx-psi — APPROVED** [review]

**Comment by mx-psi** [comment]
Needs minor edits for CI to pass but content LGTM

**Comment by braydonk** [comment]
> Needs minor edits for CI to pass

Thanks @mx-psi those are fixed now, also added something to `.yamllint` to not scan `node_modulues` when running `yamllint` locally (didn't think it was worth a whole separate PR for something that small).

File: docs/system/process-metrics.md
Comment (braydonk) [reply docs/system/process-metrics.md:150]: Hmm I'm not sure, it looks like with these `process.thread` fields in the linked docs, they all seem to be associated with a single thread itself as a resource, whereas the thread count is associated with a process resource. I worry that might be difficult to understand at first.

Regardless I will include it in the following issue for things to revisit for these metrics (kind of like #325). My plan is to open it once this is merged.

**Comment by braydonk** [comment]
Looks like these workflows require approval now if someone can press the button for me. :smile:

File: docs/system/process-metrics.md
Comment (AlexanderWert) [reply docs/system/process-metrics.md:150]: See this related issue / discussion: https://github.com/open-telemetry/semantic-conventions/issues/260

**Review by ChrsMark — APPROVED** [review]
Content-wise lgtm! 

@braydonk please file an issue for https://github.com/open-telemetry/semantic-conventions/pull/330#discussion_r1330143868 so as to examine https://github.com/open-telemetry/semantic-conventions/issues/260 as well :).

**Comment by braydonk** [comment]
Thanks, I'll file an issue today and I'll put my thoughts on #260 in there!

**Comment by joaopgrassi** [comment]
I will try to take a look at it today or tomorrow!

File: CHANGELOG.md
Comment (joaopgrassi) [inline CHANGELOG.md:12]: If you want, this can be shorter by doing the same as I did for the system metrics:

```
- Rename attributes for `system.cpu.*` metrics:
    - `state` to `system.cpu.state`
    - `cpu` to `system.cpu.logical_number`
```

File: model/metrics/process-metrics.yaml
Comment (joaopgrassi) [inline model/metrics/process-metrics.yaml:125]: In the current conventions, this is a counter.

File: schema-next.yaml
Comment (joaopgrassi) [inline schema-next.yaml:135]: I think it's better to read if the attributes listed here are split by their "namespace". Like I did for the system metrics https://github.com/open-telemetry/semantic-conventions/pull/89/files#diff-3ddd62d7281d98d63c3e92f5ae8cd7776971b44ca39f52b0140dc02363dfebda

File: docs/system/process-metrics.md
Comment (joaopgrassi) [inline docs/system/process-metrics.md:155]: We are missing `process.open_file_descriptors`

**Review by joaopgrassi — COMMENTED** [review]
Here's a first pass

File: CHANGELOG.md
Comment (braydonk) [reply CHANGELOG.md:21]: Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744) (rebased so it moved to a different place)

File: schema-next.yaml
Comment (braydonk) [reply schema-next.yaml:139]: Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744) (rebased so it moved to a different place)

File: model/metrics/process-metrics.yaml
Comment (braydonk) [reply model/metrics/process-metrics.yaml]: Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744)

File: docs/system/process-metrics.md
Comment (braydonk) [reply docs/system/process-metrics.md:157]: Resolved in [5ad3a95](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/5ad3a95ab2ddaf681fef56fbe757d1dedda68744)

File: model/metrics/process-metrics.yaml
Comment (joaopgrassi) [inline model/metrics/process-metrics.yaml:61]: The attribute must be under the namespace (`process.disk.*`) here as well

**Comment by braydonk** [comment]
It's possible I'm missing some kind of context from discussions around #51, but it seems like what I have here should be fine. With the paging fault type attribute for example: it is in the `process.paging.faults` namespace. I added that namespace, as the attribute was previously `type` (just `type`). `process.paging` is not any more of a proper namespace than `process.paging.faults` as far as I can tell.

If I'm indeed missing something extra that's not from #51, and this attribute should go to `process.paging`, then I can also rename it to `fault_type`. But I would like to understand why for this and the other pointed out attributes are not following the decision from #51 already.

**Comment by joaopgrassi** [comment]
> With the paging fault type attribute for example: it is in the `process.paging.faults` namespace

The thing I'm not sure if that `process.paging.faults` is the a metric, not a namespace. From reading #51 and the linked issues, my interpretation is that attributes should go on the namespace , meaning `process.paging.*` in this case. Maybe we need clarification if attributes can also be nested under the "implicit metric namespace". 

CC @open-telemetry/specs-semconv-approvers @open-telemetry/specs-semconv-maintainers

File: model/metrics/process-metrics.yaml
Comment (joaopgrassi) [reply model/metrics/process-metrics.yaml:128]: I opened an issue to discuss it #394. I noticed the decision is also not documented anywhere and we also have the other discussion about "metric names should be allowed to be namespaces" https://github.com/open-telemetry/semantic-conventions/issues/50

Comment (ChrsMark) [reply model/metrics/process-metrics.yaml:128]: Maybe going with sth like `process.paging.fault.count` and `process.paging.fault.type` would be more tidy?
This would also be aligned with ECS' alignment effort and not having fields conflicting with leaf nodes as described at https://github.com/open-telemetry/semantic-conventions/pull/333#issuecomment-1764384113.

**Comment by braydonk** [comment]
I have updated the PR to match the semantic convention guidance as well as I'm able.

I renamed the metrics that are `updowncounter`s to use the `.count` pluralization method. I also changed the metric attributes such that they do not use the metric names as namespaces.

File: model/metrics/process-metrics.yaml
Comment (braydonk) [reply model/metrics/process-metrics.yaml:7]: For what it's worth, I highly doubt these attributes will ever be used outside of this namespace. Is it still common practice to put these in the global registry even if it only makes sense for the attribute to belong to a single namespace?

File: docs/system/process-metrics.md
Comment (joaopgrassi) [inline docs/system/process-metrics.md:23]: Maybe consider making the section follow the same structure as in the system metrics markdown: https://github.com/braydonk/semantic-conventions/blob/process_metrics/docs/system/system-metrics.md?plain=1#L24

File: docs/system/process-metrics.md
Comment (joaopgrassi) [inline docs/system/process-metrics.md:38]: We also specify here if the metric is required/recommended https://github.com/braydonk/semantic-conventions/blob/process_metrics/docs/system/system-metrics.md?plain=1#L68

**Review by joaopgrassi — COMMENTED** [review]
I think this looks good now. Just a few markdown structural nits

File: docs/system/process-metrics.md
Comment (braydonk) [reply docs/system/process-metrics.md:38]: I think every metric in this document is going to be recommended. Should I add this to every section, or include something in the top?

**Comment by braydonk** [comment]
I added `proposed_changelog.md` and `proposed_schema_next.yaml`, which upon PR approval I will add properly to the `CHANGELOG.md` and `schema-next.yaml` files. This is to avoid constantly needing to rebase these giant files if the PR lasts over a long period of time.

File: docs/system/process-metrics.md
Comment (joaopgrassi) [reply docs/system/process-metrics.md:48]: I'd say yes, to be consistent with the other files. I have opened an issue in the past, to make this part of the markdown template in the built-tool repo.

File: model/metrics/process-metrics.yaml
Comment (ChrsMark) [reply model/metrics/process-metrics.yaml:7]: I guess for consistency reasons all attributes should go to the registry no matter how many refs exist for them.

File: docs/system/process-metrics.md
Comment (braydonk) [reply docs/system/process-metrics.md:48]: Done in [914e018](https://github.com/open-telemetry/semantic-conventions/pull/330/commits/914e018823e3bcb85fc91eddb9b948ed4d13c2ea)

**Review by joaopgrassi — APPROVED** [review]
LGTM after fixing the CI issues

File: model/metrics/process-metrics.yaml
Comment (ChrsMark) [reply model/metrics/process-metrics.yaml:7]: I'm fine with doing so. In that case let's file an issue to keep track of this unless the follow up PR comes right after :).

**Review by ChrsMark — APPROVED** [review]
LGTM. Let's make sure that we file an issue for moving the attributes to the registry: https://github.com/open-telemetry/semantic-conventions/pull/330#discussion_r1419228700

**Comment by braydonk** [comment]
> LGTM. Let's make sure that we file an issue for moving the attributes to the registry: https://github.com/open-telemetry/semantic-conventions/pull/330#discussion_r1419228700

Opened https://github.com/open-telemetry/semantic-conventions/issues/650 to track this.

**Comment by braydonk** [comment]
This should be ready for merge.

File: model/metrics/process-metrics.yaml
Comment (joaopgrassi) [reply model/metrics/process-metrics.yaml:7]: @braydonk created the issue here #650

File: docs/system/process-metrics.md
Comment (joaopgrassi) [reply docs/system/process-metrics.md]: @braydonk could you also please add an issue for this, and add to the stability system project/label `system-semconv-wg-ga-blocker`

Comment (braydonk) [reply docs/system/process-metrics.md]: I ended up changing it to `process.thread.count` in the end, so I don't think we need an issue for it.

**Review by dmitryax — APPROVED** [review]

**Comment by joaopgrassi** [comment]
I will bring this up in the semconv SIG meeting today, to raise awareness. I guess we can merge it then later today/tomorrow if no objections.

File: docs/system/process-metrics.md
Comment (joaopgrassi) [reply docs/system/process-metrics.md]: Ah right. Going to resolve this discussion then if it's fine for everyone.

**Review by AlexanderWert — APPROVED** [review]

**Review by jsuereth — APPROVED** [review]

**Pull request merged**
